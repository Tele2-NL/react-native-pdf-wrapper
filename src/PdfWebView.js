import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { NativeModules, View, Text, WebView, StyleSheet } from 'react-native';
import Pdf from 'react-native-pdf';
import RNFetchBlob from 'rn-fetch-blob';
import SHA1 from 'crypto-js/sha1';
import { FILE_TYPES, getFileType, isRequestSuccessful } from './utils';

const { PdfWrapper } = NativeModules;

const ERROR_TYPES = {
  INVALID_PDF: 'INVALID_PDF',
};

class PdfWebView extends Component {
  static propTypes = Pdf.propTypes;

  static defaultProps = {
    onError: () => {},
  };

  state = {
    isLoading: true,
    source: null,
  };

  componentDidMount() {
    this.handleShowFile();
  }

  handleError = ({ error, localPath, temporaryPath }) => {
    this.props.onError(error);

    this.handleRemoveFiles(localPath, temporaryPath);
  };

  handleCopyLocalFile = destinationPath => RNFetchBlob.fs.cp(this.props.source.uri, destinationPath);

  handleWriteBase64File = destinationPath =>
    RNFetchBlob.fs.writeFile(
      destinationPath,
      this.props.source.uri.replace(/data:application\/pdf;base64,/i, ''),
      'base64'
    );

  handleRemoveFiles = async (localPath, temporaryPath) => {
    await RNFetchBlob.fs.unlink(localPath);
    await RNFetchBlob.fs.unlink(temporaryPath);
  };

  handleRequestFile = (temporaryPath) =>
    RNFetchBlob
      .config({ path: temporaryPath })
      .fetch(
        this.props.source.method || 'GET',
        this.props.source.uri,
        this.props.source.headers || {},
        this.props.source.body || "",
      );
  
  handleSetFile = uri =>
    this.setState({
      source: {
        uri,
      },
    });

  handleShowFile = async () => {
    const { source } = this.props;

    // File will be stored under cache directory
    const localFilePath = `${RNFetchBlob.fs.dirs.CacheDir}/${SHA1(source.uri)}.pdf`;

    const fileType = getFileType(source);

    if (fileType === FILE_TYPES.LOCAL) {
      await this.handleCopyLocalFile(localFilePath);

      return this.handleSetFile(localFilePath);
    }

    if (fileType === FILE_TYPES.BASE_64) {
      await this.handleWriteBase64File(localFilePath);

      return this.handleSetFile(localFilePath);
    }

    const localTemporaryFilePath = `${localFilePath}.tmp`;

    // Remove any file with same path
    await this.handleRemoveFiles(localFilePath, localTemporaryFilePath);

    try {
      // First download to a local temporary file
      const response = await this.handleRequestFile(localTemporaryFilePath);

      // Check if request was successful
      if (!isRequestSuccessful(response)) {
        throw new Error(ERROR_TYPES.INVALID_PDF);
      }

      // Check if Pdf is not corrupted
      await PdfWrapper.isPdfValid(localTemporaryFilePath);

      // Copy the temporary file to a normal file
      await RNFetchBlob.fs.cp(localTemporaryFilePath, localFilePath)

      // Remove temporary file
      RNFetchBlob.fs.unlink(localTemporaryFilePath);

      this.handleSetFile(localFilePath);
    } catch (error) {
      console.log('NativeModules', NativeModules)
      console.log('error', error)
      this.handleError({
        error,
        localPath: localFilePath,
        temporaryPath: localTemporaryFilePath,
      });
    }
  }

  handlePdfLoaded = () => this.props.onLoadComplete();

  render() {
    const { source } = this.state;

    if (!source) {
      return null;
    }

    return (
      <WebView
        source={source}
        onLoadEnd={this.handlePdfLoaded}
        onError={e => console.log('error', e)}
      />
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
});

export default PdfWebView;
