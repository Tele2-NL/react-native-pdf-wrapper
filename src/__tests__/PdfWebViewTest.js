import React from 'react';
import { NativeModules } from 'react-native';
import { shallow } from 'enzyme';
import RNFetchBlob from 'rn-fetch-blob';
import PdfWebView from '../PdfWebView';

jest.useFakeTimers();

describe('<PdfWebView />', () => {
  beforeEach(() => {
    this.props = {
      source: {
        uri: 'https://remote/path/to/file',
      },
      onError: jest.fn(),
    };

    jest.clearAllMocks();
  });

  it('renders correctly', async () => {
    const wrapper = shallow(<PdfWebView {...this.props} />);

    jest.runAllTimers();

    expect(wrapper).toMatchSnapshot();
  });

  it('should remove previous files with the same name', () => {
    shallow(<PdfWebView {...this.props} />);

    jest.runAllTimers();

    expect(RNFetchBlob.fs.unlink).toHaveBeenCalled();
  });

  it('should check that pdf is not corrupted', () => {
    shallow(<PdfWebView {...this.props} />);

    jest.runAllTimers();

    expect(NativeModules.PdfWrapper.isPdfValid).toHaveBeenCalledTimes(1);
  });
});
