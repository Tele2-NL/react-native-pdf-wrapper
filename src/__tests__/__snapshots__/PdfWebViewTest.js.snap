// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PdfWebView /> renders correctly 1`] = `
<WebView
  onError={[MockFunction]}
  onLoadEnd={[Function]}
  originWhitelist={
    Array [
      "file://",
      "http://",
    ]
  }
  source={
    Object {
      "uri": "/path/to/cache/dir/3953e93bb2ac01a09496abbb91e2bb0070f75482.pdf",
    }
  }
/>
`;
